src/qibo/derivative.py

```diff
@@ -158,3 +158,74 @@ def circuit(nqubits = 1):
     result = float(generator_eigenval * (forward - backward) * scale_factor)
 
     return result
+
+
+def finite_differences(
+    circuit,
+    hamiltonian,
+    parameter_index,
+    initial_state=None,
+    step_size=1e-7,
+):
+    """
+    Calculate derivative of the expectation value of `hamiltonian` on the
+    final state obtained by executing `circuit` on `initial_state` with
+    respect to the variational parameter identified by `parameter_index`
+    in the circuit's parameters list. This method can be used only in
+    exact simulation mode.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit.
+        hamiltonian (:class:`qibo.hamiltonians.Hamiltonian`): target observable.
+            if you want to execute on hardware, a symbolic hamiltonian must be
+            provided as follows (example with Pauli Z and ``nqubits=1``):
+            ``SymbolicHamiltonian(np.prod([ Z(i) for i in range(1) ]))``.
+        parameter_index (int): the index which identifies the target parameter
+            in the ``circuit.get_parameters()`` list.
+        initial_state (ndarray, optional): initial state on which the circuit
+            acts. Default is ``None``.
+        step_size (float): step size used to evaluate the finite difference
+            (default 1e-7).
+
+    Returns:
+        (float): Value of the derivative of the expectation value of the hamiltonian
+            with respect to the target variational parameter.
+    """
+
+    if parameter_index > len(circuit.get_parameters()):
+        raise_error(ValueError, """This index is out of bounds.""")
+
+    if not isinstance(hamiltonian, AbstractHamiltonian):
+        raise_error(
+            TypeError,
+            "hamiltonian must be a qibo.hamiltonians.Hamiltonian or qibo.hamiltonians.SymbolicHamiltonian object",
+        )
+
+    backend = hamiltonian.backend
+
+    # parameters copies
+    parameters = np.asarray(circuit.get_parameters()).copy()
+    shifted = parameters.copy()
+
+    # shift the parameter_index element
+    shifted[parameter_index] += step_size
+    circuit.set_parameters(shifted)
+
+    # forward evaluation
+    forward = hamiltonian.expectation(
+        backend.execute_circuit(circuit=circuit, initial_state=initial_state).state()
+    )
+
+    # backward shift and evaluation
+    shifted[parameter_index] -= 2 * step_size
```

Comment:
Here I am guessing that you are doin ` - 2 * step` since before the same parameter in shifted was incremented by ` + step`. Perhaps it is better to define two shifted variables: `backward_shifted` and `forward_shifted`.

Reply:
This is useful if we want to define asymmetric step_sizes, right? I wanted to implement the standard finite difference method here, but of course we can generalize it.